docs/en/development/data_structures.md

```diff
@@ -5,4 +5,879 @@ slug: /internals
 ---
 # JuiceFS Internals
 
-Working in progress...
+## 1. Introduction
+
+This article introduces the main implementation details of JuiceFS, which is used as a reference for developers to understand and contribute open source code. The content corresponds to the JuiceFS code version v1.0.0 and the metadata version V1.
+
+## 2. Keyword Definition
+
+- File system: i.e. JuiceFS Volume, represents a separate namespace. Files can be moved freely within the same filesystem, while data copies are required between different filesystems.
+- Metadata engine: A component that stores and manages file system metadata, usually served by a database that supports transactions. There are three major categories of metadata engines that are currently supported.
+  - Redis: Redis and various protocol-compatible services
+  - SQL: MySQL, PostgreSQL, SQLite, etc.
+  - TKV: TiKV, BadgerDB, etc.
+- Datastore: A component used to store and manage file system data, usually served by object storage, such as Amazon S3, Aliyun OSS, etc. It can also be served by other storage systems that are compatible with object storage semantics, such as local file systems, Ceph Rados, TiKV, etc.
```

Comment:
```suggestion
- Datastore: is a component used to store and manage file system data, usually served by object storage, such as Amazon S3, Aliyun OSS, etc. It can also be served by other storage systems that are compatible with object storage semantics, such as local file systems, Ceph Rados, TiKV, etc.
```